Source/Charts/Renderers/BarChartRenderer.swift

```diff
@@ -241,37 +241,74 @@ open class BarChartRenderer: BarLineScatterCandleBubbleRenderer
             }
         }
         
-        let isSingleColor = dataSet.colors.count == 1
-        
-        if isSingleColor
-        {
-            context.setFillColor(dataSet.color(atIndex: 0).cgColor)
-        }
-        
         for j in buffer.indices
         {
             let barRect = buffer[j]
             
             guard viewPortHandler.isInBoundsLeft(barRect.origin.x + barRect.size.width) else { continue }
             guard viewPortHandler.isInBoundsRight(barRect.origin.x) else { break }
 
-            if !isSingleColor
-            {
-                // Set the color for the currently drawn value. If the index is out of bounds, reuse colors.
-                context.setFillColor(dataSet.color(atIndex: j).cgColor)
-            }
-            
-            context.fill(barRect)
+            drawBar(context: context, dataSet: dataSet, index: j, barRect: barRect)
             
             if drawBorder
             {
+                context.saveGState()
                 context.setStrokeColor(borderColor.cgColor)
                 context.setLineWidth(borderWidth)
                 context.stroke(barRect)
+                context.restoreGState()
             }
         }
     }
     
+    open func drawBar(context: CGContext, dataSet: BarChartDataSetProtocol, index: Int, barRect: CGRect)
+    {
+        context.saveGState()
+        
+        if let gradientColors = dataSet.barGradientColors, gradientColors.count > 0
+        {
+            if let gradientColor = dataSet.barGradientColor(atIndex: index)
```

Comment:
Why the if statement? We know `dataSet.barGradientColor(atIndex:) only returns `nil` when `dataSet.barGradientColors == nil`. It seems to me we could just write
```
if let gradientColor = dataSet.barGradientColor(atIndex: index) {
    drawGradient(context: context, barRect: barRect, gradientColors: gradientColor, orientation: dataSet.barGradientOrientation)
} // ...
```